[Pytorch Edge] Model Ops compatibility api

JacobSzwejbka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57501 [Pytorch Edge] Model Ops compatibility api**

Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

Commit:
[Pytorch Edge] Model Ops compatibility api

Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]

JacobSzwejbka commented on torch/jit/mobile/__init__.py:
```diff
@@ -1,6 +1,7 @@
 import torch
 
 from torch.jit._serialization import validate_map_location
+from typing import Dict, NamedTuple
```
oh this slipped through. TODO remove this line

kimishpatel commented on torch/csrc/jit/mobile/model_compatibility.h:
```diff
@@ -20,5 +21,20 @@ TORCH_API int64_t _get_model_bytecode_version(const std::string& filename);
     std::shared_ptr<caffe2::serialize::ReadAdapterInterface> rai);
 
+// Struct storing metadata of an operator that can be useful for versioning
+struct OperatorInfo {
```
Do you envision more stuff being added here?

JacobSzwejbka replied:
Yes. At the very least backend info will be useful from the runtimes side. Im planning on sharing this struct between runtime and model. In the future we might move to a full operator versioning system and then that would go inside this struct too.

kimishpatel commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -100,5 +102,73 @@ int64_t _get_model_bytecode_version(std::vector<IValue> bytecode_ivalues) {
+        }
+
+        // grab schema
+        int num_args = -1;
```
We should just create a sentinel values for this rather than hardcoding it in every place.

kimishpatel reviewed (changes requested):
Left a couple of comments.

raziel commented on torch/jit/mobile/__init__.py:
```diff
@@ -105,3 +106,34 @@ def _get_model_bytecode_version(f_input) -> int:
+            or a string containing a file name
+
+    Returns:
+        Operators and info: A Dictionary. A mapping of all the root ops of the
```
I think documenting the key is important.

raziel reviewed (commented):
Thanks
Some minor comments.

JacobSzwejbka commented on torch/csrc/jit/mobile/model_compatibility.cpp:
```diff
@@ -100,5 +102,73 @@ int64_t _get_model_bytecode_version(std::vector<IValue> bytecode_ivalues) {
+        if (op.size() > 2) {
+          num_args = op[2].toInt();
+        }
+        result.emplace(op_name, OperatorInfo{num_args});
```
-1 args in the schema isnt possible for a normal operator. I'm essentially using it as a null value so that this method can still be run on models from old bytecode versions and still get a useable result back. i.e. you can still at least get the model ops, but the schema information will be empty.

raziel replied:
maybe just add a comment about it

JacobSzwejbka commented on torch/csrc/jit/mobile/model_compatibility.h:
```diff
@@ -20,5 +21,20 @@ TORCH_API int64_t _get_model_bytecode_version(const std::string& filename);
+  int num_args;
+};
+
+// The family of methods below to get the root ops and information from a model
```
No by root ops its only means the top level ops. Theres no way to get what we call the 'traced' operators on a model without running the tracer. Thats the problem I was referring too in the design doc when I mentioned that getting the model ops seems difficult to me. If we want to get the full op list and not just the root ops we will need to come up with a different solution that loads the model up and traces it. cc @iseeyuan

raziel replied:
That's right. This (and the need for tracing) is the unfortunate consequence of TS IR being ignorant of the compositions done at the dispatcher level.

I suggest we document a couple of things:
1. What you mention that root ops are only provided (as you did) and that traced ops are not worth to add since it requires the current runtime as input (and consequently its current transient dependencies). 
2. The corner case for potential false positives: The model-runtime compatibility check done by validating the model-level root ops (returned by this API) are a subset of all ops (root and transient) available in a runtime (returned by the runtime API), can incorrectly pass even when the root ops in the model have transient op dependencies not present in the runtime.

I suggest we should create a "manual" type doc for these APIs that document how to use, and these caveats.

@iseeyuan

iseeyuan replied:
There is possibility of false positive on "operator subset" check. 
For other compatibility check, such as schema and versioning, it should be safer: 
If a traced op is updated, the caller ops (for example, the root ops) should be responsible for this update. So checking the root ops only would be fine. 
But for subset check there's a gap between the static model and dynamic runtime behavior. Hopefully TS IR could include those compositions including dtype, to have this check more complete.

Commit:
Update on "[Pytorch Edge] Model Ops compatibility api"


Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1693,6 +1693,15 @@ void initJitScriptBindings(PyObject* module) {
       });
+  py::class_<OperatorInfo>(m, "OperatorInfo")
+      .def_readonly("num_schema_args", &OperatorInfo::num_schema_args);
+  m.def("_get_model_ops_and_info", [](const std::string& filename) {
+    return _get_model_ops_and_info(filename);
+  });
+  m.def("_get_model_ops_and_info", [](const std::string& buffer) {
+    std::istringstream in(buffer);
+    return _get_model_ops_and_info(in);
+  });
```
The input type is the same. I'm not sure if the binding between c++ and python can work well here. Recommend to rename the one read from buffer `_get_model_ops_and_info` to `_get_model_ops_and_info_from_buffer`.

JacobSzwejbka replied:
oh good catch I meant to name it from_buffer

Commit:
Update on "[Pytorch Edge] Model Ops compatibility api"


Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM.

Commit:
Update on "[Pytorch Edge] Model Ops compatibility api"


Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]

raziel reviewed (approved):
Thanks!

Commit:
Update on "[Pytorch Edge] Model Ops compatibility api"


Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge] Model Ops compatibility api"


Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]

raziel commented on test/mobile/test_bytecode.py:
```diff
@@ -264,5 +297,12 @@ def test_backport_bytecode_from_file_to_buffer(self):
             torch.testing.assert_allclose(mobile_module_result, expected_mobile_module_result)
 
 
+    def test_get_model_ops_and_info(self):
```
can we just wait for v6 to go in? That way hardcoding this test would not be needed (the test could then be about models v6 and above to return #args > 0).

raziel commented on torch/csrc/jit/mobile/model_compatibility.h:
```diff
@@ -1,7 +1,9 @@
 #pragma once
 
+#include <torch/csrc/jit/mobile/runtime_compatibility.h>
 #include <istream>
```
```suggestion
#include <torch/csrc/jit/mobile/runtime_compatibility.h>

#include <istream>
```

raziel commented on torch/csrc/jit/mobile/runtime_compatibility.h:
```diff
@@ -1,12 +1,15 @@
 #pragma once
 
+#include <c10/util/Optional.h>
 #include <memory>
```
```suggestion
#include <c10/util/Optional.h>

#include <memory>
```

raziel reviewed (commented):
I think this should wait for v6 to go it.

@iseeyuan to sync on timeline.

JacobSzwejbka committed:
Update on "[Pytorch Edge] Model Ops compatibility api"


Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]

JacobSzwejbka committed:
Update on "[Pytorch Edge] Model Ops compatibility api"


Add an api _get_model_ops_and_info to get root operators and versioning info of a model in both cxx and python, and the input can be from a file path or buffer.

Differential Revision: [D28162765](https://our.internmc.facebook.com/intern/diff/D28162765/)

[ghstack-poisoned]
